tests/tcp_window_size_exerciser/tcp_wnd_test_server.c

```diff
@@ -0,0 +1,232 @@
+  printf("=========================================================\n\n");
+  {
+    int socketfd = 0, clientfd = 0, clientsize = 0, readsize = 0;
+    char buffer[BUFFER_SIZE] = {0};
+    struct sockaddr_in server, client;
+    int optval; /* flag value for setsockopt */
+    
+    /*Create a socket*/
+    socketfd = socket( AF_INET, SOCK_STREAM, IPPROTO_TCP);
+    if (0 > socketfd) {
+      printf("ERROR opening socket!\n");
+      exit(1);
+    }
+    printf("Socket: OK\n");
+    
+    optval = 1;
+    setsockopt(socketfd, SOL_SOCKET, SO_REUSEADDR,(const void *)&optval , sizeof(int));
+    
+    /* Built the server Internet address */
+    bzero( &server, sizeof(server));
+    server.sin_family = AF_INET;
+    server.sin_port = htons(port);
+    inet_pton( AF_INET, pServerIp, &server.sin_addr);
+      
+    if (0 != bind(socketfd, (struct sockaddr*) &server, sizeof(server))) {
+      printf("ERROR on binding!\n");
+      exit(1);
+    }
+    printf("Bind: OK\n");
+
+    if (0 > listen( socketfd, 6)) {
+      printf("ERROR on listen!\n");
+      exit(1);
+    }
+    printf("Listen: OK\n");
+  
+    clientsize = sizeof(struct sockaddr_in);
+  
+    clientfd = accept( socketfd, ( struct sockaddr*)&client, (socklen_t*)&clientsize);
+    if (0 > clientfd) {
+      printf("ERROR on accept!\n");
+      exit(1);
+    }
+    printf("Connection accepted: OK [clientfd:%d]\n", clientfd);
+
+    struct timespec start, end;
+    long long diff = 0;
+    
+    /* Set receive window size to 30k */
+    setsockopt(socketfd, SOL_SOCKET, SO_RCVBUF,(const void *)&maxwndsize, sizeof(maxwndsize));
+    setsockopt(clientfd, SOL_SOCKET, SO_RCVBUF,(const void *)&maxwndsize, sizeof(maxwndsize));
+    rcvwnd = maxwndsize;
+    clock_gettime(CLOCK_MONOTONIC, &start);	/* mark start time */
+ 
+    while (1) {
+      
+      readsize = recvfrom( clientfd, buffer, (rand() % sizeof(buffer)), 0, ( struct sockaddr*)&client, (socklen_t*)&client);
+      if (0 < readsize) {
+        printf("readsize: %d\n", readsize);
+      }
+      else {
+        printf("Something went wrong with recvfrom()! %s\n", strerror(errno));
+      }
+      
+      clock_gettime(CLOCK_MONOTONIC, &end);	/* mark the end time */
+      diff = (end.tv_sec - start.tv_sec) + (end.tv_nsec - start.tv_nsec)/1000000000;
+      if (diff > windowtimesec) {
+      	rcvwnd = rcvwnd == minwndsize ? maxwndsize : minwndsize;
+      	printf("Setsockopt: [SO_RCVBUF] window size:%d\n", rcvwnd);
+      	setsockopt(socketfd, SOL_SOCKET, SO_RCVBUF,(const void *)&rcvwnd , sizeof(rcvwnd));
+      	setsockopt(clientfd, SOL_SOCKET, SO_RCVBUF,(const void *)&rcvwnd , sizeof(rcvwnd));
+      	clock_gettime(CLOCK_MONOTONIC, &start);	/* mark start time */
+      }
+
+      if (0 < readsize) {
+        /* Buffer validation */
+        static int counter = 0;
+        {
+          int i;
+          /* message integrity validation */
+          for(i=0; i<readsize; ++i){
+            counter %= clientmsgsize;
```

Comment:
Looks to me like this will fail if the client's msgSize > 255. See comment above re. filling the clients buffer. Also, this is where the last entry in the client buffer not being filled will break as well (See another comment above).

It will be easier for you if the user writes `int`s rather than chars and fills the buffer from 0 to BUFFER_SIZE-1.